TEP: Proof of NFT Ownership (#95)
ton-blockchain/TEPs: text/0095-prove-ownership.md

##### Ownership proof
**NFT owner** can send message to NFT with this schema:
```
prove_ownership#04ded148 query_id:uint64 dest:MsgAddress 
forward_payload:^Cell with_content:Bool = InternalMsgBody;
```
After that NFT will send transfer to `dest` with scheme:
```
ownership_proof#6ecd55cc query_id:uint64 item_id:uint256 owner:MsgAddress 
data:^Cell content:(Maybe ^Cell) = InternalMsgBody;
```
If something goes wrong and target contract not accepts message, and it will be bounced back to NFT, NFT will proxy this bounce to **owner**, this way coins will not stuck on NFT.
Schema of message that will be send back to owner:
```
ownership_proof_bounced#c18e86d2 query_id:uint64 item_id:uint256 owner:MsgAddress 
data:^Cell content:(Maybe ^Cell) = InternalMsgBody;
```

##### Owner info
**anyone** can send message to NFT with this schema:
```
request_owner#d0c3bfea query_id:uint64 dest:MsgAddress 
forward_payload:^Cell with_content:Bool = InternalMsgBody;
```
After that NFT will send transfer to `dest` with scheme:
```
owner_info#c2fa9387 query_id:uint64 item_id:uint256 initiator:MsgAddress owner:MsgAddress 
data:^Cell content:(Maybe ^Cell) = InternalMsgBody;
```

> If something goes wrong and target contract not accepts message, and it will be bounced back to NFT, NFT will proxy this bounce to **initiator**.

Reviewer: ```suggestion
If the target contract fails to process the message, it will be bounced back to NFT, NFT then should proxy that bounced message to the **initiator**.
```

> If something goes wrong and target contract not accepts message, and it will be bounced back to NFT, NFT will proxy this bounce to **initiator**.

Reviewer: Should the NFT check that `initiator` != `dest`? Imagine a (supposedly) common case when a contract requests a proof for itself and does not process the request properly. Seems like it will get spammed with a bounce. Would we avoid an infinite back-and-forth in such case?
